python/bandpassUtils.py

```diff
@@ -0,0 +1,251 @@
+    """
+    # Read the filter files.
+    filterfiles = glob(os.path.join(filterDir, '*-bandResponse.dat'))
+    filters = {}
+    for f in filterfiles:
+        fname = os.path.split(f)[1].split('-')[0]
+        filters[fname] = Bandpass()
+        filters[fname].readThroughput(f)
+    if addLosses:
+        # Read and multiply the losses.
+        lossfiles = glob(os.path.join(filterDir, 'filterLosses', '*.dat'))
+        loss = Bandpass()
+        loss.readThroughputList(lossfiles)
+        for f in filters:
+            wavelen, sb = filters[f].multiplyThroughputs(loss.wavelen, loss.sb)
+            filters[f].setBandpass(wavelen, sb)
+    return filters
+
+def savitzky_golay(y, window_size=31, order=3, deriv=0, rate=1):
+    # y = throughput for lenses
+    try:
+        window_size = np.abs(np.int(window_size))
+        order = np.abs(np.int(order))
+    except ValueError, msg:
+        raise ValueError("window_size and order have to be of type int")
+    if window_size % 2 != 1 or window_size < 1:
+        raise TypeError("window_size size must be a positive odd number")
+    if window_size < order + 2:
+        raise TypeError("window_size is too small for the polynomials order")
+    order_range = range(order+1)
+    half_window = (window_size -1) // 2
+    # precompute coefficients
+    b = np.asmatrix([[k**i for i in order_range] for k in range(-half_window, half_window+1)])
+    m = np.linalg.pinv(b).A[deriv] * rate**deriv * np.math.factorial(deriv)
+    # pad the signal at the extremes with
+    # values taken from the signal itself
+    firstvals = y[0] - np.abs(y[1:half_window+1][::-1] - y[0] )
+    lastvals = y[-1] + np.abs(y[-half_window-1:-1][::-1] - y[-1])
+    y = np.concatenate((firstvals, y, lastvals))
+    return np.convolve(m[::-1], y, mode='valid')
+
+def buildLens(lensDir, addLosses=True):
+    """
+    Build each individual lens. Assumes there are files l*Glass.dat, l*BBAR.dat, and a subdirectory with loss files.
+    Returns a bandpass object.
+    """
+    lens = Bandpass()
+    # Read the glass base file.
+    glassfile = glob(os.path.join(lensDir, 'l*Glass.dat'))
+    if len(glassfile) != 1:
+        raise ValueError('Expected a single glass file in this directory, found: ', glassfile)
+    glassfile = glassfile[0]
+    glass = Bandpass()
+    glass.readThroughput(glassfile)
+    # Smooth the glass response.
+    smoothGlassSb = savitzky_golay(glass.sb, 31, 3)
+    lens.setBandpass(glass.wavelen, smoothGlassSb)
+    # Read the broad band antireflective (BBAR) coatings files.
+    bbarfiles = glob(os.path.join(lensDir, 'l*BBAR.dat'))
+    bbars = Bandpass()
+    bbars.readThroughputList(bbarfiles)
+    # Multiply the bbars by the glass.
+    wavelen, sb = lens.multiplyThroughputs(bbars.wavelen, bbars.sb)
+    lens.setBandpass(wavelen, sb)
+    # Add losses.
+    if addLosses:
+        lossfiles = glob(os.path.join(lensDir, 'l*Losses", ''*.dat'))
```

Comment:
It did somehow (results of the bandpass look like the version calculated by Chuck, so must include losses), but I will fix.